lib/framedivider.rb

```diff
@@ -390,46 +422,284 @@ def opening(model, id)
+    # If all else fails, TBD will calculate an approximate whole product
+    # U-factor by adding up the subsurface's construction material thermal
+    # resistances (as well as the subsurface's parent surface film resistances).
+    # This is the least accurate option, especially if subsurfaces have Frame
+    # & Divider objects.
+
+    u = s.uFactor
+    u = s.additionalProperties.getFeatureAsDouble("uFactor") if u.empty?
+    u = c.additionalProperties.getFeatureAsDouble("uFactor") if u.empty?
+    if u.empty?
+      r = rsi(c, surface.filmResistance)
+      if r < TOL
+        msg = "Subsurface '#{id}' U-factor unavailable - skipping"
+        TBD.log(TBD::ERROR, msg)
+        next
+      else
+        u = 1.0 / r
+      end
+    else
+      u = u.get
+    end
+
+    # Should verify convexity of vertex wire/face ...
+    #
+    #       A
+    #      / \
+    #     /   \
+    #    /     \
+    #   / C --- D    <<< allowed as OpenStudio/E+ subsurface?
+    #  / /
+    #  B
+    #
+    # Should convert (annoying) 4-point subsurface into triangle ...
+    #        A
+    #       / \
+    #      /   \
+    #     /     \
+    #    B - C - D   <<< allowed as OpenStudio/E+ subsurface?
+    #
+    four = true if s.vertices.size == 4
+
+    if s.windowPropertyFrameAndDivider.empty?
+      vec = s.vertices
+      area = gross
+    else
+      width = s.windowPropertyFrameAndDivider.get.frameWidth
+      ptz = offset(s.vertices, width)
+      fd = true
+
+      # Re-convert Topolys 3D points into OpenStudio 3D points.
+      vec = OpenStudio::Point3dVector.new
+      vec << OpenStudio::Point3d.new(ptz[:A][:p].x, ptz[:A][:p].y, ptz[:A][:p].z)
+      vec << OpenStudio::Point3d.new(ptz[:B][:p].x, ptz[:B][:p].y, ptz[:B][:p].z)
+      vec << OpenStudio::Point3d.new(ptz[:C][:p].x, ptz[:C][:p].y, ptz[:C][:p].z)
+      vec << OpenStudio::Point3d.new(ptz[:D][:p].x, ptz[:D][:p].y, ptz[:D][:p].z) if four
+
+      tr1 = []
+      tr1 << ptz[:A][:p]
+      tr1 << ptz[:B][:p]
+      tr1 << ptz[:C][:p]
+      area1 = areaHeron(tr1)
+
+      area2 = 0
+      if four
+        tr2 = []
+        tr2 << ptz[:A][:p]
+        tr2 << ptz[:C][:p]
+        tr2 << ptz[:D][:p]
+        area2 = areaHeron(tr2)
+      end
+      area = area1 + area2
+    end
+
+    sub = { v:      s.vertices,
+            points: vec,
+            n:      n,
+            gross:  gross,
+            area:   area,
+            type:   type,
+            u:      u }
+
+    sub[:glazed] = true if glazed
+    subs[id] = sub
+  end
+
+  # Test for conflicts (with fits?, overlaps?) between surfaces to determine
+  # whether to keep original points or switch to std::vector of revised
+  # coordinates, offset by Frame & Divider frame width. This will also
+  # inadvertently catch pre-existing (yet nonetheless invalid) OpenStudio inputs
+  # (without Frame & Dividers).
+  valid = true
+  subs.each do |id, sub|
+    next unless fd
+    next unless valid
+    unless fits?(surface.vertices, sub[:points])
```

Comment:
Wouldn't you want to make `fits?` directional?  You don't want a base surface that fits inside the sub surface?